x-pack/platform/plugins/shared/alerting/server/application/rule/methods/bulk_fill_gaps_by_rule_ids/bulk_fill_gaps_by_rule_ids.ts

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import pMap from 'p-map';
+import Boom from '@hapi/boom';
+import { groupBy, mapValues } from 'lodash';
+import type { RulesClientContext } from '../../../../rules_client';
+import type {
+  BulkFillGapsByRuleIdsResult,
+  BulkFillGapsByRuleIdsParams,
+  BulkFillGapsByRuleIdsOptions,
+} from './types';
+import { logProcessedAsAuditEvent, toBulkGapFillError } from './utils';
+import { AlertingAuthorizationEntity, WriteOperations } from '../../../../authorization';
+import { batchBackfillRuleGaps } from './batch_backfill_rule_gaps';
+import { bulkFillGapsByRuleIdParamsSchema } from './schemas';
+
+const DEFAULT_MAX_BACKFILL_CONCURRENCY = 10;
+
+export const bulkFillGapsByRuleIds = async (
+  context: RulesClientContext,
+  { rules, range }: BulkFillGapsByRuleIdsParams,
+  options: BulkFillGapsByRuleIdsOptions
+): Promise<BulkFillGapsByRuleIdsResult> => {
+  const result = bulkFillGapsByRuleIdParamsSchema.safeParse(range);
+  if (!result.success) {
+    const message = result.error.errors.map((error) => error.message).join(' ');
+    throw Boom.badRequest(
+      `Error validating backfill schedule parameters "${JSON.stringify(range)}" - ${message}`
+    );
+  }
+
+  const errored: BulkFillGapsByRuleIdsResult['errored'] = [];
+  const skipped: BulkFillGapsByRuleIdsResult['skipped'] = [];
+  const backfilled: BulkFillGapsByRuleIdsParams['rules'] = [];
+  const eventLogClient = await context.getEventLogClient();
+  const maxBackfillConcurrency =
+    options?.maxBackfillConcurrency ?? DEFAULT_MAX_BACKFILL_CONCURRENCY;
+
+  // Make sure user has access to these rules
+  const rulesByAlertType = mapValues(
+    groupBy(rules, (rule) => `${rule.alertTypeId}<>${rule.consumer}`),
+    (groupedRules) => ({
+      alertTypeId: groupedRules[0].alertTypeId,
```

Comment:
```suggestion
      ruleTypeId: groupedRules[0].alertTypeId,
```

I know the field is called `alertTypeId` but it's a legacy naming that is difficult to change since it's a saved object field. In our code, we try to use the correct terminology 🙈

Reply:
Done